File: cxx/src/index_v2/snapshot_format.cpp
#include "index_v2/snapshot_format.h"

#include <algorithm>
#include <cassert>
#include <istream>
#include <ostream>
#include <stdexcept>
#include <string>

namespace swiftlist::index_v2 {

namespace {

// Maximum string length read from snapshot to prevent large allocation on
// corrupt/malicious input (64 MiB).
inline constexpr int32_t kMaxStringLength = 64 * 1024 * 1024;

void Write7BitEncoded(std::ostream& s, int32_t value) {
    uint32_t v = static_cast<uint32_t>(value);
    while (v >= 0x80) {
        s.put(static_cast<char>((v & 0x7Fu) | 0x80u));
        v >>= 7;
    }
    s.put(static_cast<char>(v));
}

int32_t Read7BitEncoded(std::istream& s) {
    int32_t result = 0;
    int shift = 0;
    char b;
    do {
        b = static_cast<char>(s.get());
        if (!s) throw std::runtime_error("Unexpected EOF reading 7-bit encoded int");
        result |= static_cast<int32_t>(static_cast<uint8_t>(b) & 0x7Fu) << shift;
        shift += 7;
    } while (static_cast<uint8_t>(b) & 0x80);
    return result;
}

void WriteString(std::ostream& s, const std::string& str) {
    Write7BitEncoded(s, static_cast<int32_t>(str.size()));
    s.write(str.data(), static_cast<std::streamsize>(str.size()));
}

std::string ReadString(std::istream& s) {
    auto len = Read7BitEncoded(s);
    if (len < 0) throw std::runtime_error("Negative string length in snapshot");
    if (len > kMaxStringLength)
        throw std::runtime_error("String length exceeds maximum in snapshot");
    std::string result(static_cast<size_t>(len), '\0');
    s.read(result.data(), len);
    if (!s) throw std::runtime_error("Unexpected EOF reading string");
    return result;
}

void WriteUInt128(std::ostream& s, const UInt128& v) {
    // Little-endian: low first, then high (matches C# BinaryWriter).
    s.write(reinterpret_cast<const char*>(&v.low), sizeof(v.low));
    s.write(reinterpret_cast<const char*>(&v.high), sizeof(v.high));
}

UInt128 ReadUInt128(std::istream& s) {
    UInt128 v;
    s.read(reinterpret_cast<char*>(&v.low), sizeof(v.low));
    s.read(reinterpret_cast<char*>(&v.high), sizeof(v.high));
    if (!s) throw std::runtime_error("Unexpected EOF reading UInt128");
    return v;
}

void WriteBytes(std::ostream& s, const void* data, std::streamsize len) {
    s.write(static_cast<const char*>(data), len);
}

void AlignStream(std::ostream& s, int alignment) {
    auto pos = s.tellp();
    auto rem = static_cast<int>(pos % alignment);
    if (rem != 0) {
        int pad = alignment - rem;
        for (int i = 0; i < pad; ++i) s.put('\0');
    }
}

} // namespace

Meta ReadHeader(std::istream& stream) {
    Meta meta;

    uint64_t magic;
    stream.read(reinterpret_cast<char*>(&magic), sizeof(magic));
    if (magic != kMagic)
        throw std::runtime_error("Snapshot magic mismatch");

    int32_t version;
    stream.read(reinterpret_cast<char*>(&version), sizeof(version));
    if (version != kVersion)
        throw std::runtime_error("Snapshot version mismatch");

    auto readInt = [&]() -> int {
        int32_t v;
        stream.read(reinterpret_cast<char*>(&v), sizeof(v));
        if (!stream) throw std::runtime_error("Unexpected EOF reading int32");
        return static_cast<int>(v);
    };

    meta.RowCount = readInt();
    meta.UniqueCount = readInt();
    meta.NameBlobLength = readInt();
    meta.ChildrenLength = readInt();
    meta.AliasEntryCount = readInt();
    meta.AliasBlobLength = readInt();
    meta.OrphanCount = readInt();
    meta.TotalFiles = readInt();
    meta.TotalDirs = readInt();

    uint8_t sourceKind = 0, idKind = 0;
    stream.read(reinterpret_cast<char*>(&sourceKind), 1);
    stream.read(reinterpret_cast<char*>(&idKind), 1);
    meta.Source = static_cast<SourceKind>(sourceKind);
    meta.IdKind = static_cast<IdKind>(idKind);

    stream.read(reinterpret_cast<char*>(&meta.VolumeSerialNumber),
               sizeof(meta.VolumeSerialNumber));
    meta.RootId = ReadUInt128(stream);
    stream.read(reinterpret_cast<char*>(&meta.JournalId), sizeof(meta.JournalId));
    stream.read(reinterpret_cast<char*>(&meta.NextUsn), sizeof(meta.NextUsn));

    meta.SourceKey = ReadString(stream);
    meta.SourceRoot = ReadString(stream);
    meta.FileSystemType = ReadString(stream);

    uint8_t complete = 0;
    stream.read(reinterpret_cast<char*>(&complete), 1);
    meta.IsComplete = (complete != 0);

    meta.ExclusionRulesFingerprint = ReadString(stream);
    meta.AliasProvidersFingerprint = ReadString(stream);
    stream.read(reinterpret_cast<char*>(&meta.LastUpdated), sizeof(meta.LastUpdated));

    return meta;
}

void WriteHeader(std::ostream& stream, const Meta& meta) {
    WriteBytes(stream, &kMagic, sizeof(kMagic));

    int32_t version = kVersion;
    WriteBytes(stream, &version, sizeof(version));

    auto writeInt = [&](int v) {
        int32_t sv = static_cast<int32_t>(v);
        WriteBytes(stream, &sv, sizeof(sv));
    };

    writeInt(meta.RowCount);
    writeInt(meta.UniqueCount);
    writeInt(meta.NameBlobLength);
    writeInt(meta.ChildrenLength);
    writeInt(meta.AliasEntryCount);
    writeInt(meta.AliasBlobLength);
    writeInt(meta.OrphanCount);
    writeInt(meta.TotalFiles);
    writeInt(meta.TotalDirs);

    uint8_t sourceKind = static_cast<uint8_t>(meta.Source);
    uint8_t idKind = static_cast<uint8_t>(meta.IdKind);
    stream.write(reinterpret_cast<const char*>(&sourceKind), 1);
    stream.write(reinterpret_cast<const char*>(&idKind), 1);

    WriteBytes(stream, &meta.VolumeSerialNumber, sizeof(meta.VolumeSerialNumber));
    WriteUInt128(stream, meta.RootId);
    WriteBytes(stream, &meta.JournalId, sizeof(meta.JournalId));
    WriteBytes(stream, &meta.NextUsn, sizeof(meta.NextUsn));

    WriteString(stream, meta.SourceKey);
    WriteString(stream, meta.SourceRoot);
    WriteString(stream, meta.FileSystemType);

    uint8_t complete = meta.IsComplete ? 1 : 0;
    stream.write(reinterpret_cast<const char*>(&complete), 1);

    WriteString(stream, meta.ExclusionRulesFingerprint);
    WriteString(stream, meta.AliasProvidersFingerprint);
    WriteBytes(stream, &meta.LastUpdated, sizeof(meta.LastUpdated));
}

std::vector<int64_t> ComputeSectionOffsets(const Meta& meta, int64_t* totalLength) {
    const int N = static_cast<int>(SnapshotSection::Count);
    const int row = meta.RowCount;
    const int uniq = meta.UniqueCount;

    // Byte size of each section (before alignment padding).
    int64_t sizes[] = {
        4LL * row,                              // NameIds
        2LL * row,                              // Flags
        4LL * row,                              // ParentIndexes
        8LL * uniq,                             // UniqueMasks
        4LL * (uniq + 1LL),                     // NameOffsets
        static_cast<int64_t>(meta.NameBlobLength), // NameBlob
        16LL * row,                             // Ids
        8LL * row,                              // Sizes
        4LL * row,                              // CreationTimes
        4LL * row,                              // LastWriteTimes
        4LL * row,                              // LastAccessTimes
        4LL * (row + 1LL),                      // ChildStarts
        4LL * meta.ChildrenLength,              // Children
        4LL * (uniq + 1LL),                     // UidStarts
        4LL * row,                              // UidRows
        4LL * (uniq + 1LL),                     // AliasStarts
        4LL * (meta.AliasEntryCount + 1LL),     // AliasEntryOffsets
        static_cast<int64_t>(meta.AliasEntryCount), // AliasProviderIds
        static_cast<int64_t>(meta.AliasBlobLength), // AliasBlob
        4LL * meta.OrphanCount,                 // OrphanRows
        16LL * meta.OrphanCount,                // OrphanFrns
        8LL * ((uniq + 63LL) / 64LL),           // UniqueAsciiBits
    };
    static_assert(N == std::size(sizes), "section count mismatch");

    // Align header end to 16 bytes (mirrors C# behavior).
    // We don't know the header size exactly here; caller handles pre-section alignment.
    std::vector<int64_t> offsets(N);

    if (totalLength) {
        int64_t acc = 0;
        for (int i = 0; i < N; ++i) {
            // Align start of each section.
            int64_t rem = acc % kSectionAlignment;
            if (rem != 0) acc += (kSectionAlignment - rem);
            offsets[i] = acc;
            acc += sizes[i];
        }
        // Align end to 16 bytes.
        int64_t rem = acc % kSectionAlignment;
        if (rem != 0) acc += (kSectionAlignment - rem);
        *totalLength = acc;
    }
    return offsets;
}

} // namespace swiftlist::index_v2

File: Core/IndexV2/Persistence/SnapshotFormat.cs
using System.Text;

namespace SwiftList.Core.IndexV2.Persistence;

internal enum SnapshotSection
{
    NameIds, Flags, ParentIndexes, UniqueMasks, NameOffsets, NameBlob,
    Ids, Sizes, CreationTimes, LastWriteTimes, LastAccessTimes,
    ChildStarts, Children, UidStarts, UidRows,
    AliasStarts, AliasEntryOffsets, AliasProviderIds, AliasBlob,
    OrphanRows, OrphanFrns,
    // One bit per unique name: 1 = pure ASCII, so its raw UTF-8 bytes ARE its chars and the search
    // hot path can match them with zero decode (FzfBytePattern). Baked at write time instead of
    // re-probing per candidate per query.
    UniqueAsciiBits,
}

// Single source of truth for the V2 snapshot layout: one memory-mappable file whose columnar
// sections ARE the runtime layout. The variable-length meta block sits after the fixed header;
// sections start at Meta.SectionsOffset, each aligned so a typed span over any section (including
// the 16-byte UInt128 id columns) is naturally aligned.
internal static class SnapshotFormat
{
    public const ulong Magic = 0x0000005844494C53; // "SLIDX\0\0\0" little-endian
    // Bumped 8 -> 9: Sector-aligned read for non-resident $ATTRIBUTE_LIST entries and full MFT extent scanning.
    public const int Version = 9;
    public const int SectionAlignment = 16;

    internal sealed class Meta
    {
        public string SourceKey = string.Empty;
        // Fully-formed root prefix ("C:\", "\\server\share\", ...) -- stored, not derived, so UNC and
        // WSL sources round-trip without the reader guessing.
        public string SourceRoot = string.Empty;
        public FileRecordSourceKind SourceKind;
        public FileRecordIdKind IdKind;
        public string FileSystemType = string.Empty;
        public uint VolumeSerialNumber;
        public UInt128 RootId;
        public ulong JournalId;
        public long NextUsn;
        public int RowCount;
        public int UniqueCount;
        public int NameBlobLength;
        public int ChildrenLength;
        public int AliasEntryCount;
        public int AliasBlobLength;
        public int OrphanCount;
        public int TotalFiles;
        public int TotalDirs;
        // Meaningless for local USN/MFT drives (always true/empty there -- see FileRecordStore.
        // IsComplete/ExclusionRulesFingerprint); load-bearing for network/WSL/folder-index drives'
        // resumable-scan diffing (TreeDiffBaseline) and exclusion-rule-change detection.
        public bool IsComplete;
        public string ExclusionRulesFingerprint = string.Empty;
        // AliasProviderRegistry.ComputeProvidersFingerprint() as of the write that produced this
        // snapshot's alias data. Compared against the CURRENT set of installed alias providers on load
        // (see UsnIndexerCacheExtensions/NetworkDriveCacheLocator) -- a mismatch means a provider was
        // installed/removed or bumped its own Version since, so every unique name's aliases need
        // regenerating via a forced recompaction.
        public string AliasProvidersFingerprint = string.Empty;
        // Wall-clock time this snapshot's data was last current -- FileRecordStore.LastUpdated's V2
        // equivalent, surfaced to the UI (e.g. "last refreshed" for a network drive).
        public DateTime LastUpdated;
        public long SectionsOffset;
    }

    public static void WriteHeader(BinaryWriter writer, Meta meta)
    {
        writer.Write(Magic);
        writer.Write(Version);
        writer.Write(meta.RowCount);
        writer.Write(meta.UniqueCount);
        writer.Write(meta.NameBlobLength);
        writer.Write(meta.ChildrenLength);
        writer.Write(meta.AliasEntryCount);
        writer.Write(meta.AliasBlobLength);
        writer.Write(meta.OrphanCount);
        writer.Write(meta.TotalFiles);
        writer.Write(meta.TotalDirs);
        writer.Write((byte)meta.SourceKind);
        writer.Write((byte)meta.IdKind);
        writer.Write(meta.VolumeSerialNumber);
        writer.Write((ulong)meta.RootId);
        writer.Write((ulong)(meta.RootId >> 64));
        writer.Write(meta.JournalId);
        writer.Write(meta.NextUsn);
        writer.Write(meta.SourceKey);
        writer.Write(meta.SourceRoot);
        writer.Write(meta.FileSystemType);
        writer.Write(meta.IsComplete);
        writer.Write(meta.ExclusionRulesFingerprint);
        writer.Write(meta.AliasProvidersFingerprint);
        writer.Write(meta.LastUpdated.ToUniversalTime().Ticks);
        // SectionsOffset is derived, not stored: the reader recomputes it from its own stream position
        // after the variable-length strings, then aligns -- see FinishHeader/ReadHeader staying in sync.
    }

    // Pads the stream from the end of the header to the aligned section base and records it in meta.
    public static void FinishHeader(Stream stream, Meta meta)
    {
        meta.SectionsOffset = Align(stream.Position);
        while (stream.Position < meta.SectionsOffset)
            stream.WriteByte(0);
    }

    public static Meta ReadHeader(BinaryReader reader)
    {
        if (reader.ReadUInt64() != Magic)
            throw new InvalidDataException("IndexV2 snapshot magic mismatch");
        if (reader.ReadInt32() != Version)
            throw new InvalidDataException("IndexV2 snapshot version mismatch");

        var meta = new Meta
        {
            RowCount = reader.ReadInt32(),
            UniqueCount = reader.ReadInt32(),
            NameBlobLength = reader.ReadInt32(),
            ChildrenLength = reader.ReadInt32(),
            AliasEntryCount = reader.ReadInt32(),
            AliasBlobLength = reader.ReadInt32(),
            OrphanCount = reader.ReadInt32(),
            TotalFiles = reader.ReadInt32(),
            TotalDirs = reader.ReadInt32(),
            SourceKind = (FileRecordSourceKind)reader.ReadByte(),
            IdKind = (FileRecordIdKind)reader.ReadByte(),
            VolumeSerialNumber = reader.ReadUInt32(),
        };
        var rootLow = reader.ReadUInt64();
        var rootHigh = reader.ReadUInt64();
        meta.RootId = new UInt128(rootHigh, rootLow);
        meta.JournalId = reader.ReadUInt64();
        meta.NextUsn = reader.ReadInt64();
        meta.SourceKey = reader.ReadString();
        meta.SourceRoot = reader.ReadString();
        meta.FileSystemType = reader.ReadString();
        meta.IsComplete = reader.ReadBoolean();
        meta.ExclusionRulesFingerprint = reader.ReadString();
        meta.AliasProvidersFingerprint = reader.ReadString();
        meta.LastUpdated = new DateTime(reader.ReadInt64(), DateTimeKind.Utc).ToLocalTime();
        meta.SectionsOffset = Align(reader.BaseStream.Position);
        return meta;
    }

    // Section offsets derive from meta counts alone, in declaration order, each 16-aligned -- the
    // writer seeks to these positions and the reader maps spans over them, so this method is the only
    // place layout order exists.
    public static long[] ComputeSectionOffsets(Meta meta, out long totalLength)
    {
        var sizes = new long[]
        {
            4L * meta.RowCount,            // NameIds
            2L * meta.RowCount,            // Flags
            4L * meta.RowCount,            // ParentIndexes
            8L * meta.UniqueCount,         // UniqueMasks
            4L * (meta.UniqueCount + 1),   // NameOffsets
            meta.NameBlobLength,           // NameBlob
            16L * meta.RowCount,           // Ids
            8L * meta.RowCount,            // Sizes
            4L * meta.RowCount,            // CreationTimes
            4L * meta.RowCount,            // LastWriteTimes
            4L * meta.RowCount,            // LastAccessTimes
            4L * (meta.RowCount + 1),      // ChildStarts
            4L * meta.ChildrenLength,      // Children
            4L * (meta.UniqueCount + 1),   // UidStarts
            4L * meta.RowCount,            // UidRows
            4L * (meta.UniqueCount + 1),   // AliasStarts
            4L * (meta.AliasEntryCount + 1), // AliasEntryOffsets
            meta.AliasEntryCount,          // AliasProviderIds
            meta.AliasBlobLength,          // AliasBlob
            4L * meta.OrphanCount,         // OrphanRows
            16L * meta.OrphanCount,        // OrphanFrns
            8L * ((meta.UniqueCount + 63) / 64), // UniqueAsciiBits
        };

        var offsets = new long[sizes.Length];
        var cursor = meta.SectionsOffset;
        for (var i = 0; i < sizes.Length; i++)
        {
            cursor = Align(cursor);
            offsets[i] = cursor;
            cursor += sizes[i];
        }
        totalLength = Align(cursor);
        return offsets;
    }

    // Reads just the small fixed+variable header (no mmap) -- for cache-discovery scans that need a
    // summary of every drive's snapshot without paying for a full mapping of each one.
    public static Meta? TryReadHeaderFromFile(string path)
    {
        try
        {
            using var stream = File.OpenRead(path);
            using var reader = new BinaryReader(stream, NameEncoding);
            return ReadHeader(reader);
        }
        catch (IOException)
        {
            throw; // transient (e.g. mid-write) -- propagate so callers don't mistake this for corruption
        }
        catch (Exception ex)
        {
            Logger.Log($"[SnapshotFormat] Failed to read header {path}: {ex.Message}", LogLevel.Error);
            return null;
        }
    }

    public static long Align(long offset) => (offset + SectionAlignment - 1) & ~(long)(SectionAlignment - 1);

    // The lookup dictionaries live only for the duration of a build; UTF-8 encoding is fixed.
    public static readonly Encoding NameEncoding = Encoding.UTF8;
}
